### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by JuanGQCadavid**
The user's project started with its endpoints and project structure.

### CODE REVIEW

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package com.ecommercepractice.userservice.models;
+
+import lombok.*;
+
+@Getter @Setter @NoArgsConstructor @AllArgsConstructor @ToString
+public class User {
+    private  String user_id;
+    private  String first_name;
+    private  String second_name;
+    private  boolean is_active;
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: change to camelCase format

File: user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
+package com.ecommercepractice.userservice.service;
+
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java]: remove space

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @DeleteMapping(path = "{user_id}")
+    public void deleteUser(@PathVariable("user_id") String user_id){
+        log.info("USER | DELETE | USER_ID  { " + user_id + "}");
+        userService.deleteUser(user_id);
+    }
+
+
+    @PutMapping(path = "{user_id}")
+    public void updateUser(@PathVariable("user_id") String user_id, @RequestBody User user){
+        log.info("USER | PUT | USER_ID  { " + user_id +  "  - USER {" + user.toString() + "}" + "}");
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: please use String.format

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    //api/v1/user/:user_name
+    @GetMapping(path = "{user_id}")
+    public User getUserByUserName(@PathVariable("user_id") String user_id){
+        log.info("USER | GET | USER_ID  { " + user_id + "}");
+        return userService.getUserbyUserName(user_id)
+                .orElse(null);
+    }
+    @DeleteMapping(path = "{user_id}")
+    public void deleteUser(@PathVariable("user_id") String user_id){
+        log.info("USER | DELETE | USER_ID  { " + user_id + "}");
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: String.format

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    @GetMapping
+    public List<User> getAllUser(){
+        log.info("USER | GET ALL");
+        return userService.getAllusers();
+    }
+
+    //api/v1/user/:user_name
+    @GetMapping(path = "{user_id}")
+    public User getUserByUserName(@PathVariable("user_id") String user_id){
+        log.info("USER | GET | USER_ID  { " + user_id + "}");
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: String.format

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+    @Override
+    public Optional<User> selectUserByUserName(String user_name) {
+        return DB.stream()
+                .filter(user -> user.getUser_id().equals(user_name))
+                .findFirst();
+    }
+
+    @Override
+    public int updateUserbyUserName(String user_name, User user) {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java]: use camel case format

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        return 1;
+    }
+
+    @Override
+    public List<User> selectAllUsers() {
+        return DB;
+    }
+
+    @Override
+    public Optional<User> selectUserByUserName(String user_name) {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java]: use camel case format

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                        return 1;
+                    }
+                    else{
+                        return 0;
+                    }
+                }).orElse(0);
+    }
+
+    @Override
+    public int deleteUserbyUserName(String user_name) {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java]: use camel case format

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/IUserDao.java [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.util.List;
+import java.util.Optional;
+
+public interface IUserDao {
+
+    int insertUser(User user);
+    List<User> selectAllUsers();
+
+    Optional<User> selectUserByUserName(String user_name);
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/IUserDao.java]: use camel case format

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java [hunk 10]
```diff
@@ -0,0 +1,13 @@
+package com.ecommercepractice.userservice.models;
+
+import lombok.*;
+
+@Getter @Setter @NoArgsConstructor @AllArgsConstructor @ToString
+public class User {
+    private  String user_id;
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: use camel case format on all the attributes

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.ecommercepractice.userservice.models;
+
+import lombok.*;
+
+@Getter @Setter @NoArgsConstructor @AllArgsConstructor @ToString
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: please use a line per @

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package com.ecommercepractice.userservice.models;
+
+import lombok.*;
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: avoid wildcard import

File: user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+@Service
+public class UserService {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java:14]: please use camel case format on this class

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RequestMapping("api/v1/user")
+@RestController
+@Slf4j // Lombok
+public class UserControlle {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: Fix Class name to UserController

Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java (hunk 13 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java:14]: Ok

Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java:14]: Carlos, its a class definition, so it should be UserService

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java (hunk 12 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java (hunk 11 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: OK

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java (hunk 10 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/IUserDao.java (hunk 9 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/dao/IUserDao.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java (hunk 5 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java (hunk 4 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java (hunk 3 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserControlle.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java (hunk 2 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/service/UserService.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java (hunk 1 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RequestMapping("api/v1/user")
+@RestController
+@Slf4j // Lombok
+public class UserController {
+    private final UserService userService;
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Please annotate userService with @Autowired and remove this
 @Autowired
    public UserController(UserService userService) {
        this.userService = userService;
    }

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Autowired
+    public UserController(UserService userService) {
+        this.userService = userService;
+    }
+
+    @PostMapping
+    public void addUser(@RequestBody User user){
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: a Rest Service always must return something, in this case, should return a HttpStatus.CREATED code

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+        log.info(String.format("USER | POST {%s}", user.toString()));
+        userService.addUser(user);
+    }
+
+    @GetMapping
+    public List<User> getAllUser(){
+        log.info("USER | GET ALL");
+        return userService.getAllusers();
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Use ResponseEntity Class with the correct HttpCode Status

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    //api/v1/user/:user_name
+    @GetMapping(path = "{userId}")
+    public User getUserByUserName(@PathVariable("userId") String userId){
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: USe ResponseEntity Class with the correct HttpCode Status

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    }
+    @DeleteMapping(path = "{userId}")
+    public void deleteUser(@PathVariable("userId") String userId){
+        log.info(String.format("USER | DELETE | USER_ID  {%s}",userId));
+        userService.deleteUser(userId);
+    }
+
+
+    @PutMapping(path = "{userId}")
+    public void updateUser(@PathVariable("userId") String userId, @RequestBody User user){
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: USe ResponseEntity Class with the correct HttpCode Status

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+        return 1;
+    }
+
+    @Override
+    public List<User> selectAllUsers() {
+        return DB;
+    }
+
+    @Override
+    public Optional<User> selectUserByUserName(String userId) {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java]: Should not return an Optional, should Return a User o throws an Exception indicating that the information was not found

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 18 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 17 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 16 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 15 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Ok

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java (hunk 20 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/dao/FakeUserDataAccessService.java]: I propose DAO returns Optional and Service handle the exception

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+@RestController
+@Slf4j // Lombok
+public class UserController {
+    @Autowired
+    private UserService userService;
+
+    @Autowired
+    private UserModelAssembler userModelAssembler;
+
+    public UserController() {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Delete this constructor

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+    public HttpHeaders connectionWithOtherService(){
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Custom-Header", "foo");
+        return headers;
+    }
+
+
+    @PostMapping
+    public ResponseEntity<EntityModel<User>> addUser(@Valid @RequestBody User user){
+
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java:51]: Delete all spaces.

File: user-service/src/main/java/com/ecommercepractice/userservice/exception/ControllerAdvisor.java [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.stream.Collectors;
+
+@ControllerAdvice
+public class ControllerAdvisor extends ResponseEntityExceptionHandler {
+
+    @ExceptionHandler(UserAlreadyCreatedException.class)
+    public ResponseEntity<Object> handleUserAlreadyCreatedException(
+            UserAlreadyCreatedException ex, WebRequest request){
+
+        JSONObject body = new JSONObject();
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/exception/ControllerAdvisor.java]: Update JSONObjet to you own class for error details

**Review by czuluagah — COMMENTED** [review]
Juan you are missing the following topics:

Swagger
Spring Data
Documentation on every class file.
JUnit Test
Error Mapping

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 21 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 22 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java:51]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/exception/ControllerAdvisor.java (hunk 23 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/exception/ControllerAdvisor.java]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+import com.ecommercepractice.userservice.models.User;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.boot.CommandLineRunner;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.Bean;
+import springfox.documentation.swagger2.annotations.EnableSwagger2;
+
+@SpringBootApplication
+@EnableSwagger2
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java]: Remove @EnableSwagger2

File: user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+
+			// fetch all customers
+			log.info("Customers found with findAll():");
+			log.info("-------------------------------");
+			for (User user : repository.findAll()) {
+				log.info(user.toString());
+			}
+			log.info("");
+
+			/*
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java]: Remove all the code that is not used anymore

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+    public HttpHeaders connectionWithOtherService(){
+        HttpHeaders headers = new HttpHeaders();
+        headers.add("Custom-Header", "foo");
+        return headers;
+    }
+
+    @ApiOperation(value = "CREATE a user.", response = User.class, responseContainer = "EntityModel")
+    @PostMapping
+    public ResponseEntity<EntityModel<User>> addUser(
+            @ApiParam(name="UserData",
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: use the same name of the param

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/H2UserDataAccessService.java [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+import com.ecommercepractice.userservice.models.User;
+import com.ecommercepractice.userservice.repository.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Optional;
+
+@Repository("h2DAO")
+public class H2UserDataAccessService implements IUserDao {
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/dao/H2UserDataAccessService.java]: This class should end in Repository instead Service

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java [hunk 28]
```diff
@@ -0,0 +1,59 @@
+package com.ecommercepractice.userservice.models;
+
+import io.swagger.annotations.ApiModel;
+import io.swagger.annotations.ApiModelProperty;
+import lombok.*;
```
Comment (czuluagah) [inline user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: don use wildcard imports

File: user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java (hunk 24 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java (hunk 25 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/UserServiceApplication.java]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java (hunk 26 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/controller/UserController.java]: done

File: user-service/src/main/java/com/ecommercepractice/userservice/dao/H2UserDataAccessService.java (hunk 27 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/dao/H2UserDataAccessService.java]: Done

File: user-service/src/main/java/com/ecommercepractice/userservice/models/User.java (hunk 28 above)
Comment (JuanGQCadavid) [reply user-service/src/main/java/com/ecommercepractice/userservice/models/User.java]: Done

**Pull request merged**